cmake/external/abseil-cpp.cmake

```diff
@@ -50,96 +50,98 @@ endif()
 # TODO: since multiple ORT's dependencies depend on Abseil, the list below would vary from version to version.
 # We'd better to not manually manage the list.
 set(ABSEIL_LIBS
+absl::absl_check
 absl::absl_log
-absl::log_internal_log_impl
-absl::log_internal_strip
-absl::log_internal_message
-absl::log_internal_format
-absl::synchronization
-absl::str_format
-absl::flags
-absl::log_internal_globals
-absl::kernel_timeout_internal
-absl::str_format_internal
-absl::hash
-absl::log_internal_append_truncated
 absl::absl_vlog_is_on
-absl::flags_commandlineflag
-absl::time
-absl::symbolize
-absl::graphcycles_internal
-absl::log_internal_conditions
-absl::strings
-absl::malloc_internal
-absl::demangle_internal
-absl::optional
-absl::stacktrace
-absl::base
-absl::demangle_rust
+absl::algorithm_container
+absl::atomic_hook
 absl::bad_optional_access
-absl::strings_internal
-absl::debugging_internal
-absl::int128
-absl::spinlock_wait
-absl::decode_rust_punycode
-absl::raw_logging_internal
-absl::flat_hash_set
-absl::flat_hash_map
-absl::node_hash_map
-absl::node_hash_set
-absl::compare
+absl::base
 absl::base_internal
-absl::nullability
-absl::bounded_utf8_length_sequence
-absl::log_severity
-absl::type_traits
-absl::atomic_hook
 absl::bits
-absl::flags_commandlineflag_internal
-absl::hash_container_defaults
-absl::numeric_representation
-absl::node_slot_policy
+absl::bounded_utf8_length_sequence
+absl::charset
+absl::city
+absl::civil_time
+absl::compare
+absl::config
+absl::container_memory
 absl::core_headers
+absl::debugging_internal
+absl::decode_rust_punycode
+absl::demangle_internal
+absl::demangle_rust
 absl::dynamic_annotations
-absl::utf8_for_code_point
+absl::endian
 absl::errno_saver
-absl::absl_check
-absl::hash_function_defaults
-absl::function_ref
-absl::city
-absl::low_level_hash
+absl::examine_stack
+absl::fast_type_id
 absl::fixed_array
-absl::variant
-absl::meta
-absl::log_internal_voidify
-absl::log_sink
-absl::log_internal_log_sink_set
-absl::log_sink_registry
+absl::flags
+absl::flags_commandlineflag
+absl::flags_commandlineflag_internal
+absl::flags_config
+absl::flags_internal
+absl::flags_reflection
+absl::flat_hash_map
+absl::flat_hash_set
+absl::function_ref
+absl::graphcycles_internal
+absl::hash
+absl::hash_container_defaults
+absl::hash_function_defaults
+absl::inlined_vector
+absl::int128
+absl::kernel_timeout_internal
 absl::log_entry
 absl::log_globals
+absl::log_internal_append_truncated
+absl::log_internal_check_op
```

Comment:
added `absl::log_internal_check_op`, other changes are just sorting the targets alphabetically.

Reply:
Because they have dependency relationship with each other, we cannot sort them alphabetically. We need a topo sort.  I wrote a script for doing this.  Ideally we should be able to extract the dependency relationship  from cmake's database instead of manually managing them. 

```python
import graphlib

graph = {
    "absl::stacktrace": {
        "absl::debugging_internal",
        "absl::config",
        "absl::core_headers",
        "absl::dynamic_annotations",
        "absl::raw_logging_internal",
    },
    "absl::errno_saver": {
        "absl::config"
    },
    "absl::debugging_internal": {
        "absl::core_headers",
        "absl::config",
        "absl::dynamic_annotations",
        "absl::errno_saver",
        "absl::raw_logging_internal",
    },
    "absl::malloc_internal": {
        "absl::base",
        "absl::base_internal",
        "absl::config",
        "absl::core_headers",
        "absl::dynamic_annotations",
        "absl::raw_logging_internal"
    },
    "absl::demangle_internal": {
        "absl::config",
        "absl::demangle_rust"
    },
    "absl::demangle_rust": {
    "absl::config",
"absl::core_headers",
"absl::decode_rust_punycode"
    },
    "absl::decode_rust_punycode":{
    "absl::bounded_utf8_length_sequence",
"absl::config",
"absl::nullability",
"absl::utf8_for_code_point"
    },
    "absl::utf8_for_code_point":{
      "absl::config"
    },
    "absl::bits": {
      "absl::core_headers",
    },
    "absl::bounded_utf8_length_sequence": {
    "absl::bits",
"absl::config"
    },
    "absl::symbolize": {
        "absl::debugging_internal",
        "absl::demangle_internal",
        "absl::base",
        "absl::config",
        "absl::core_headers",
        "absl::dynamic_annotations",
        "absl::malloc_internal",
        "absl::raw_logging_internal",
        "absl::strings",
    },
    "absl::graphcycles_internal": {
        "absl::base",
        "absl::base_internal",
        "absl::config",
        "absl::core_headers",
        "absl::malloc_internal",
        "absl::raw_logging_internal",
    },
    "absl::kernel_timeout_internal": {
        "absl::base",
        "absl::config",
        "absl::core_headers",
        "absl::raw_logging_internal",
        "absl::time",
    },
    "absl::synchronization": {
        "absl::graphcycles_internal",
        "absl::kernel_timeout_internal",
        "absl::atomic_hook",
        "absl::base",
        "absl::base_internal",
        "absl::config",
        "absl::core_headers",
        "absl::dynamic_annotations",
        "absl::malloc_internal",
        "absl::raw_logging_internal",
        "absl::stacktrace",
        "absl::symbolize",
        "absl::time",
    },
    "absl::nullability": {"absl::config", "absl::core_headers", "absl::type_traits"},
    "absl::dynamic_annotations": {"absl::config"},
    "absl::atomic_hook": {"absl::config", "absl::core_headers"},
    "absl::raw_logging_internal": {
        "absl::atomic_hook",
        "absl::config",
        "absl::core_headers",
        "absl::errno_saver",
        "absl::log_severity",
    },
    "absl::base_internal": {"absl::config", "absl::type_traits"},
    "absl::base": {
        "absl::atomic_hook",
        "absl::base_internal",
        "absl::config",
        "absl::core_headers",
        "absl::dynamic_annotations",
        "absl::log_severity",
        "absl::nullability",
        "absl::raw_logging_internal",
        "absl::spinlock_wait",
        "absl::type_traits",
    },
    "absl::spinlock_wait":{"absl::base_internal",
"absl::core_headers",
"absl::errno_saver"},
    "absl::type_traits": {"absl::config", "absl::core_headers"},
    "absl::compare": {"absl::config", "absl::core_headers", "absl::type_traits"},
    "absl::int128": {
        "absl::compare",
        "absl::config",
        "absl::core_headers",
        "absl::bits",
    },
    "absl::strings_internal":{
        "absl::config",
        "absl::core_headers",
        "absl::endian",
        "absl::raw_logging_internal",
        "absl::type_traits"
    },
    "absl::strings":{
    "absl::string_view",
"absl::strings_internal",
"absl::base",
"absl::bits",
"absl::charset",
"absl::config",
"absl::core_headers",
"absl::endian",
"absl::int128",
"absl::memory",
"absl::nullability",
"absl::raw_logging_internal",
"absl::throw_delegate",
"absl::type_traits"
    },
    "absl::time": {
        "absl::base",
        "absl::civil_time",
        "absl::core_headers",
        "absl::int128",
        "absl::raw_logging_internal",
        "absl::strings",
        "absl::time_zone",
    },
    "absl::bad_optional_access":{
        "absl::config",
        "absl::raw_logging_internal",
    },
    "absl::optional": {
        "absl::bad_optional_access",
        "absl::base_internal",
        "absl::config",
        "absl::core_headers",
        "absl::memory",
        "absl::nullability",
        "absl::type_traits",
        "absl::utility"
    },
    "absl::log_severity": {"absl::config", "absl::core_headers"},
    "absl::flags_commandlineflag": {
        "absl::config",
        "absl::fast_type_id",
        "absl::flags_commandlineflag_internal",
        "absl::optional",
        "absl::strings",
    },
    "absl::flags_commandlineflag_internal": {
        "absl::config",
        "absl::dynamic_annotations",
        "absl::fast_type_id",
    },
    "absl::core_headers": {"absl::config"},
    "absl::flags": {
        "absl::config",
        "absl::flags_commandlineflag",
        "absl::flags_config",
        "absl::flags_internal",
        "absl::flags_reflection",
        "absl::core_headers",
        "absl::strings",
    },
    "absl::absl_log": {"absl::log_internal_log_impl"},
    "absl::log_internal_log_impl": {
        "absl::log_internal_conditions",
        "absl::log_internal_message",
        "absl::log_internal_strip",
        "absl::absl_vlog_is_on",
    },
    "absl::absl_vlog_is_on": {
        "absl::vlog_config_internal",
        "absl::config",
        "absl::core_headers",
        "absl::strings",
    },
    "absl::log_internal_strip": {
        "absl::core_headers",
        "absl::log_internal_message",
        "absl::log_internal_nullstream",
        "absl::log_severity",
    },
    "absl::log_internal_append_truncated": {
        "absl::config",
        "absl::strings",
        "absl::span"
    },
    "absl::node_slot_policy":{
      "absl::config"
    },
    "absl::node_hash_set": {
        "absl::container_memory",
        "absl::core_headers",
        "absl::hash_container_defaults",
        "absl::node_slot_policy",
        "absl::raw_hash_set",
        "absl::algorithm_container",
        "absl::memory",
        "absl::type_traits",
    },
    "absl::node_hash_map": {
      "absl::container_memory",
"absl::core_headers",
"absl::hash_container_defaults",
"absl::node_slot_policy",
"absl::raw_hash_map",
"absl::algorithm_container",
"absl::memory",
"absl::type_traits"
    },
    "absl::log_internal_format": {
    "absl::config",
"absl::core_headers",
"absl::log_internal_append_truncated",
"absl::log_internal_config",
"absl::log_internal_globals",
"absl::log_severity",
"absl::strings",
"absl::str_format",
"absl::time",
"absl::span",
    },
    "absl::log_internal_globals":{
        "absl::config",
        "absl::core_headers",
        "absl::log_severity",
        "absl::raw_logging_internal",
        "absl::strings",
        "absl::time"
    },
    "absl::log_internal_message": {
        "absl::base",
        "absl::config",
        "absl::core_headers",
        "absl::errno_saver",
        "absl::inlined_vector",
        "absl::examine_stack",
        "absl::log_internal_append_truncated",
        "absl::log_internal_format",
        "absl::log_internal_globals",
        "absl::log_internal_proto",
        "absl::log_internal_log_sink_set",
        "absl::log_internal_nullguard",
        "absl::log_globals",
        "absl::log_entry",
        "absl::log_severity",
        "absl::log_sink",
        "absl::log_sink_registry",
        "absl::memory",
        "absl::raw_logging_internal",
        "absl::strings",
        "absl::strerror",
        "absl::time",
        "absl::span",
    },
    "absl::log_internal_conditions": {
        "absl::base",
        "absl::config",
        "absl::core_headers",
        "absl::log_internal_voidify",
    },
    "absl::hash":{
    "absl::bits",
"absl::city",
"absl::config",
"absl::core_headers",
"absl::endian",
"absl::fixed_array",
"absl::function_ref",
"absl::meta",
"absl::int128",
"absl::strings",
"absl::optional",
"absl::variant",
"absl::utility",
"absl::low_level_hash"
    },
    "absl::flat_hash_map":{
        "absl::container_memory",
        "absl::core_headers",
        "absl::hash_container_defaults",
        "absl::raw_hash_map",
        "absl::algorithm_container",
        "absl::type_traits"
    },
    "absl::hash_container_defaults": {
    "absl::config",
"absl::hash_function_defaults"
    },
    "absl::flat_hash_set": {
        "absl::container_memory",
        "absl::hash_container_defaults",
        "absl::raw_hash_set",
        "absl::algorithm_container",
        "absl::core_headers",
        "absl::memory",
        "absl::type_traits"
    },
    "re2::re2": {
        "absl::absl_check",
        "absl::absl_log",
        "absl::base",
        "absl::core_headers",
        "absl::fixed_array",
        "absl::flags",
        "absl::flat_hash_map",
        "absl::flat_hash_set",
        "absl::hash",
        "absl::inlined_vector",
        "absl::optional",
        "absl::span",
        "absl::str_format",
        "absl::strings",
        "absl::synchronization",
    },
    "absl::str_format":{
        "absl::config",
        "absl::core_headers",
        "absl::nullability",
        "absl::span",
        "absl::str_format_internal",
        "absl::string_view"
    },
    "absl::str_format_internal": {
        "absl::bits",
        "absl::strings",
        "absl::config",
        "absl::core_headers",
        "absl::fixed_array",
        "absl::inlined_vector",
        "absl::numeric_representation",
        "absl::type_traits",
        "absl::utility",
        "absl::int128",
        "absl::span"
    },
    "absl::numeric_representation": {
      "absl::config"
    },
    "absl::log_internal_message": {
        "absl::base",
        "absl::config",
        "absl::core_headers",
        "absl::errno_saver",
        "absl::inlined_vector",
        "absl::examine_stack",
        "absl::log_internal_append_truncated",
        "absl::log_internal_format",
        "absl::log_internal_globals",
        "absl::log_internal_proto",
        "absl::log_internal_log_sink_set",
        "absl::log_internal_nullguard",
        "absl::log_globals",
        "absl::log_entry",
        "absl::log_severity",
        "absl::log_sink",
        "absl::log_sink_registry",
        "absl::memory",
        "absl::raw_logging_internal",
        "absl::strings",
        "absl::strerror",
        "absl::time",
        "absl::span",
    },
}
ts = graphlib.TopologicalSorter(graph)
for e in reversed(tuple(ts.static_order())):
  print(e)
```

Reply:
I added something that can find the sub dependencies of a target automatically, but it's only used when assembling the static libraries for the apple framework. 

https://github.com/microsoft/onnxruntime/blob/25b1c38e87c6a976849d127eca7e00c8d9956619/cmake/onnxruntime.cmake#L380

Idea was from https://github.com/google/dawn/blob/main/src/cmake/BundleLibraries.cmake.

Something similar _could_ be used to automatically iterate the higher level absl dependencies.

Reply:
Then this list should not be so long, and we should not need to specify any internal static lib.  So, what went wrong?

Reply:
RE2 depends on 
```
        "absl::absl_check",
        "absl::absl_log",
        "absl::base",
        "absl::core_headers",
        "absl::fixed_array",
        "absl::flags",
        "absl::flat_hash_map",
        "absl::flat_hash_set",
        "absl::hash",
        "absl::inlined_vector",
        "absl::optional",
        "absl::span",
        "absl::str_format",
        "absl::strings",
        "absl::synchronization",
```
They are almost all we need.

Reply:
I'm not sure why I was seeing the errors. possibly from some intermediate commit. anyway, I can't reproduce the error after reverting those changes so I'll consider them unnecessary.

Reply:
FWIW, that library is included in the link command after I reverted the ABSEIL_LIBS changes. search for libabsl_log_internal_check_op.a.

```
Ld /Users/edch2/source/onnxruntime/build/ios/Debug/Debug-iphonesimulator/onnxruntime_test_all.app/PlugIns/onnxruntime_test_all_xc.xctest/onnxruntime_test_all_xc normal (in target 'onnxruntime_test_all_xc' from project 'onnxruntime')
    cd /Users/edch2/source/onnxruntime/cmake
    /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/clang++ -Xlinker -reproducible -target x86_64-apple-ios13.0-simulator -bundle -isysroot /Applications/Xcode.app/Contents/Developer/Platforms/iPhoneSimulator.platform/Developer/SDKs/iPhoneSimulator18.0.sdk -O0 -L/Users/edch2/source/onnxruntime/build/ios/Debug/build/EagerLinkingTBDs/Debug-iphonesimulator -L/Users/edch2/source/onnxruntime/build/ios/Debug/Debug-iphonesimulator -L/Applications/Xcode.app/Contents/Developer/Platforms/iPhoneSimulator.platform/Developer/usr/lib -F/Users/edch2/source/onnxruntime/build/ios/Debug/build/EagerLinkingTBDs/Debug-iphonesimulator -F/Users/edch2/source/onnxruntime/build/ios/Debug/Debug-iphonesimulator -F/Applications/Xcode.app/Contents/Developer/Platforms/iPhoneSimulator.platform/Developer/Library/Frameworks -iframework /Applications/Xcode.app/Contents/Developer/Platforms/iPhoneSimulator.platform/Developer/Library/Frameworks -iframework /Applications/Xcode.app/Contents/Developer/Platforms/iPhoneSimulator.platform/Developer/SDKs/iPhoneSimulator18.0.sdk/Developer/Library/Frameworks -filelist /Users/edch2/source/onnxruntime/build/ios/Debug/build/onnxruntime_test_all_xc.build/Debug-iphonesimulator/Objects-normal/x86_64/onnxruntime_test_all_xc.LinkFileList -Xlinker -rpath -Xlinker /Applications/Xcode.app/Contents/Developer/Platforms/iPhoneSimulator.platform/Developer/Library/Frameworks -Xlinker -rpath -Xlinker @loader_path/Frameworks -Xlinker -rpath -Xlinker @executable_path/Frameworks -dead_strip -bundle_loader /Users/edch2/source/onnxruntime/build/ios/Debug/Debug-iphonesimulator/onnxruntime_test_all.app/onnxruntime_test_all -Xlinker -object_path_lto -Xlinker /Users/edch2/source/onnxruntime/build/ios/Debug/build/onnxruntime_test_all_xc.build/Debug-iphonesimulator/Objects-normal/x86_64/onnxruntime_test_all_xc_lto.o -Xlinker -no_deduplicate -Xlinker -objc_abi_version -Xlinker 2 -fobjc-arc -fobjc-link-runtime -bundle -Wl,-headerpad_max_install_names -framework Foundation -framework XCTest /Users/edch2/source/onnxruntime/build/ios/Debug/Debug-iphonesimulator/libonnxruntime_test_utils.a /Users/edch2/source/onnxruntime/build/ios/Debug/Debug-iphonesimulator/libonnxruntime_session.a /Users/edch2/source/onnxruntime/build/ios/Debug/Debug-iphonesimulator/libonnxruntime_optimizer.a /Users/edch2/source/onnxruntime/build/ios/Debug/Debug-iphonesimulator/libonnxruntime_providers.a /Users/edch2/source/onnxruntime/build/ios/Debug/Debug-iphonesimulator/libonnxruntime_util.a /Users/edch2/source/onnxruntime/build/ios/Debug/Debug-iphonesimulator/libonnxruntime_lora.a /Users/edch2/source/onnxruntime/build/ios/Debug/Debug-iphonesimulator/libonnxruntime_framework.a /Users/edch2/source/onnxruntime/build/ios/Debug/Debug-iphonesimulator/libonnxruntime_graph.a /Users/edch2/source/onnxruntime/build/ios/Debug/Debug-iphonesimulator/libonnxruntime_mlas.a /Users/edch2/source/onnxruntime/build/ios/Debug/Debug-iphonesimulator/libonnxruntime_common.a /Users/edch2/source/onnxruntime/build/ios/Debug/Debug-iphonesimulator/libonnxruntime_flatbuffers.a /Users/edch2/source/onnxruntime/build/ios/Debug/Debug-iphonesimulator/libonnx_test_data_proto.a /Users/edch2/source/onnxruntime/build/ios/Debug/lib/Debug/libgtest.a /Users/edch2/source/onnxruntime/build/ios/Debug/lib/Debug/libgmock.a /Users/edch2/source/onnxruntime/build/ios/Debug/_deps/onnx-build/Debug-iphonesimulator/libonnx.a /Users/edch2/source/onnxruntime/build/ios/Debug/_deps/onnx-build/Debug-iphonesimulator/libonnx_proto.a /Users/edch2/source/onnxruntime/build/ios/Debug/_deps/protobuf-build/Debug-iphonesimulator/libprotobuf-lited.a /Users/edch2/source/onnxruntime/build/ios/Debug/_deps/re2-build/Debug-iphonesimulator/libre2.a /Users/edch2/source/onnxruntime/build/ios/Debug/_deps/flatbuffers-build/Debug-iphonesimulator/libflatbuffers.a /Users/edch2/source/onnxruntime/build/ios/Debug/_deps/abseil_cpp-build/absl/log/Debug-iphonesimulator/libabsl_log_internal_message.a /Users/edch2/source/onnxruntime/build/ios/Debug/_deps/abseil_cpp-build/absl/log/Debug-iphonesimulator/libabsl_log_internal_format.a /Users/edch2/source/onnxruntime/build/ios/Debug/_deps/abseil_cpp-build/absl/synchronization/Debug-iphonesimulator/libabsl_synchronization.a /Users/edch2/source/onnxruntime/build/ios/Debug/_deps/abseil_cpp-build/absl/log/Debug-iphonesimulator/libabsl_log_internal_globals.a /Users/edch2/source/onnxruntime/build/ios/Debug/_deps/abseil_cpp-build/absl/synchronization/Debug-iphonesimulator/libabsl_kernel_timeout_internal.a /Users/edch2/source/onnxruntime/build/ios/Debug/_deps/abseil_cpp-build/absl/strings/Debug-iphonesimulator/libabsl_str_format_internal.a /Users/edch2/source/onnxruntime/build/ios/Debug/_deps/abseil_cpp-build/absl/hash/Debug-iphonesimulator/libabsl_hash.a /Users/edch2/source/onnxruntime/build/ios/Debug/_deps/abseil_cpp-build/absl/flags/Debug-iphonesimulator/libabsl_flags_commandlineflag.a /Users/edch2/source/onnxruntime/build/ios/Debug/_deps/abseil_cpp-build/absl/time/Debug-iphonesimulator/libabsl_time.a /Users/edch2/source/onnxruntime/build/ios/Debug/_deps/abseil_cpp-build/absl/debugging/Debug-iphonesimulator/libabsl_symbolize.a /Users/edch2/source/onnxruntime/build/ios/Debug/_deps/abseil_cpp-build/absl/synchronization/Debug-iphonesimulator/libabsl_graphcycles_internal.a /Users/edch2/source/onnxruntime/build/ios/Debug/_deps/abseil_cpp-build/absl/log/Debug-iphonesimulator/libabsl_log_internal_conditions.a /Users/edch2/source/onnxruntime/build/ios/Debug/_deps/abseil_cpp-build/absl/strings/Debug-iphonesimulator/libabsl_strings.a /Users/edch2/source/onnxruntime/build/ios/Debug/_deps/abseil_cpp-build/absl/base/Debug-iphonesimulator/libabsl_malloc_internal.a /Users/edch2/source/onnxruntime/build/ios/Debug/_deps/abseil_cpp-build/absl/debugging/Debug-iphonesimulator/libabsl_demangle_internal.a /Users/edch2/source/onnxruntime/build/ios/Debug/_deps/abseil_cpp-build/absl/debugging/Debug-iphonesimulator/libabsl_stacktrace.a /Users/edch2/source/onnxruntime/build/ios/Debug/_deps/abseil_cpp-build/absl/base/Debug-iphonesimulator/libabsl_base.a /Users/edch2/source/onnxruntime/build/ios/Debug/_deps/abseil_cpp-build/absl/debugging/Debug-iphonesimulator/libabsl_demangle_rust.a /Users/edch2/source/onnxruntime/build/ios/Debug/_deps/abseil_cpp-build/absl/types/Debug-iphonesimulator/libabsl_bad_optional_access.a /Users/edch2/source/onnxruntime/build/ios/Debug/_deps/abseil_cpp-build/absl/strings/Debug-iphonesimulator/libabsl_strings_internal.a /Users/edch2/source/onnxruntime/build/ios/Debug/_deps/abseil_cpp-build/absl/debugging/Debug-iphonesimulator/libabsl_debugging_internal.a /Users/edch2/source/onnxruntime/build/ios/Debug/_deps/abseil_cpp-build/absl/numeric/Debug-iphonesimulator/libabsl_int128.a /Users/edch2/source/onnxruntime/build/ios/Debug/_deps/abseil_cpp-build/absl/base/Debug-iphonesimulator/libabsl_spinlock_wait.a /Users/edch2/source/onnxruntime/build/ios/Debug/_deps/abseil_cpp-build/absl/debugging/Debug-iphonesimulator/libabsl_decode_rust_punycode.a /Users/edch2/source/onnxruntime/build/ios/Debug/_deps/abseil_cpp-build/absl/base/Debug-iphonesimulator/libabsl_raw_logging_internal.a /Users/edch2/source/onnxruntime/build/ios/Debug/_deps/abseil_cpp-build/absl/base/Debug-iphonesimulator/libabsl_log_severity.a /Users/edch2/source/onnxruntime/build/ios/Debug/_deps/abseil_cpp-build/absl/flags/Debug-iphonesimulator/libabsl_flags_commandlineflag_internal.a /Users/edch2/source/onnxruntime/build/ios/Debug/_deps/abseil_cpp-build/absl/debugging/Debug-iphonesimulator/libabsl_utf8_for_code_point.a /Users/edch2/source/onnxruntime/build/ios/Debug/_deps/abseil_cpp-build/absl/hash/Debug-iphonesimulator/libabsl_city.a /Users/edch2/source/onnxruntime/build/ios/Debug/_deps/abseil_cpp-build/absl/hash/Debug-iphonesimulator/libabsl_low_level_hash.a /Users/edch2/source/onnxruntime/build/ios/Debug/_deps/abseil_cpp-build/absl/log/Debug-iphonesimulator/libabsl_log_sink.a /Users/edch2/source/onnxruntime/build/ios/Debug/_deps/abseil_cpp-build/absl/log/Debug-iphonesimulator/libabsl_log_internal_log_sink_set.a /Users/edch2/source/onnxruntime/build/ios/Debug/_deps/abseil_cpp-build/absl/log/Debug-iphonesimulator/libabsl_log_entry.a /Users/edch2/source/onnxruntime/build/ios/Debug/_deps/abseil_cpp-build/absl/log/Debug-iphonesimulator/libabsl_log_globals.a /Users/edch2/source/onnxruntime/build/ios/Debug/_deps/abseil_cpp-build/absl/log/Debug-iphonesimulator/libabsl_log_internal_nullguard.a /Users/edch2/source/onnxruntime/build/ios/Debug/_deps/abseil_cpp-build/absl/debugging/Debug-iphonesimulator/libabsl_examine_stack.a /Users/edch2/source/onnxruntime/build/ios/Debug/_deps/abseil_cpp-build/absl/log/Debug-iphonesimulator/libabsl_log_internal_proto.a /Users/edch2/source/onnxruntime/build/ios/Debug/_deps/abseil_cpp-build/absl/base/Debug-iphonesimulator/libabsl_strerror.a /Users/edch2/source/onnxruntime/build/ios/Debug/_deps/abseil_cpp-build/absl/container/Debug-iphonesimulator/libabsl_raw_hash_set.a /Users/edch2/source/onnxruntime/build/ios/Debug/_deps/abseil_cpp-build/absl/log/Debug-iphonesimulator/libabsl_vlog_config_internal.a /Users/edch2/source/onnxruntime/build/ios/Debug/_deps/abseil_cpp-build/absl/flags/Debug-iphonesimulator/libabsl_flags_reflection.a /Users/edch2/source/onnxruntime/build/ios/Debug/_deps/abseil_cpp-build/absl/flags/Debug-iphonesimulator/libabsl_flags_internal.a /Users/edch2/source/onnxruntime/build/ios/Debug/_deps/abseil_cpp-build/absl/flags/Debug-iphonesimulator/libabsl_flags_config.a /Users/edch2/source/onnxruntime/build/ios/Debug/_deps/abseil_cpp-build/absl/time/Debug-iphonesimulator/libabsl_time_zone.a /Users/edch2/source/onnxruntime/build/ios/Debug/_deps/abseil_cpp-build/absl/time/Debug-iphonesimulator/libabsl_civil_time.a /Users/edch2/source/onnxruntime/build/ios/Debug/_deps/abseil_cpp-build/absl/strings/Debug-iphonesimulator/libabsl_string_view.a /Users/edch2/source/onnxruntime/build/ios/Debug/_deps/abseil_cpp-build/absl/base/Debug-iphonesimulator/libabsl_throw_delegate.a /Users/edch2/source/onnxruntime/build/ios/Debug/_deps/pytorch_clog-build/Debug-iphonesimulator/libclog.a /Users/edch2/source/onnxruntime/build/ios/Debug/_deps/google_nsync-build/Debug-iphonesimulator/libnsync_cpp.a /Users/edch2/source/onnxruntime/build/ios/Debug/_deps/pytorch_cpuinfo-build/Debug-iphonesimulator/libcpuinfo.a /Users/edch2/source/onnxruntime/build/ios/Debug/_deps/pytorch_clog-build/Debug-iphonesimulator/libclog.a -framework Foundation /Users/edch2/source/onnxruntime/build/ios/Debug/lib/Debug/libgtest.a /Users/edch2/source/onnxruntime/build/ios/Debug/_deps/abseil_cpp-build/absl/flags/Debug-iphonesimulator/libabsl_flags_private_handle_accessor.a /Users/edch2/source/onnxruntime/build/ios/Debug/_deps/abseil_cpp-build/absl/flags/Debug-iphonesimulator/libabsl_flags_commandlineflag.a /Users/edch2/source/onnxruntime/build/ios/Debug/_deps/abseil_cpp-build/absl/flags/Debug-iphonesimulator/libabsl_flags_commandlineflag_internal.a /Users/edch2/source/onnxruntime/build/ios/Debug/_deps/abseil_cpp-build/absl/flags/Debug-iphonesimulator/libabsl_flags_marshalling.a /Users/edch2/source/onnxruntime/build/ios/Debug/_deps/abseil_cpp-build/absl/flags/Debug-iphonesimulator/libabsl_flags_program_name.a /Users/edch2/source/onnxruntime/build/ios/Debug/_deps/abseil_cpp-build/absl/strings/Debug-iphonesimulator/libabsl_cord.a /Users/edch2/source/onnxruntime/build/ios/Debug/_deps/abseil_cpp-build/absl/strings/Debug-iphonesimulator/libabsl_cordz_info.a /Users/edch2/source/onnxruntime/build/ios/Debug/_deps/abseil_cpp-build/absl/strings/Debug-iphonesimulator/libabsl_cord_internal.a /Users/edch2/source/onnxruntime/build/ios/Debug/_deps/abseil_cpp-build/absl/strings/Debug-iphonesimulator/libabsl_cordz_functions.a /Users/edch2/source/onnxruntime/build/ios/Debug/_deps/abseil_cpp-build/absl/strings/Debug-iphonesimulator/libabsl_cordz_handle.a /Users/edch2/source/onnxruntime/build/ios/Debug/_deps/abseil_cpp-build/absl/crc/Debug-iphonesimulator/libabsl_crc_cord_state.a /Users/edch2/source/onnxruntime/build/ios/Debug/_deps/abseil_cpp-build/absl/crc/Debug-iphonesimulator/libabsl_crc32c.a /Users/edch2/source/onnxruntime/build/ios/Debug/_deps/abseil_cpp-build/absl/crc/Debug-iphonesimulator/libabsl_crc_internal.a /Users/edch2/source/onnxruntime/build/ios/Debug/_deps/abseil_cpp-build/absl/crc/Debug-iphonesimulator/libabsl_crc_cpu_detect.a /Users/edch2/source/onnxruntime/build/ios/Debug/_deps/abseil_cpp-build/absl/container/Debug-iphonesimulator/libabsl_raw_hash_set.a /Users/edch2/source/onnxruntime/build/ios/Debug/_deps/abseil_cpp-build/absl/container/Debug-iphonesimulator/libabsl_hashtablez_sampler.a /Users/edch2/source/onnxruntime/build/ios/Debug/_deps/abseil_cpp-build/absl/profiling/Debug-iphonesimulator/libabsl_exponential_biased.a /Users/edch2/source/onnxruntime/build/ios/Debug/_deps/abseil_cpp-build/absl/log/Debug-iphonesimulator/libabsl_log_internal_conditions.a /Users/edch2/source/onnxruntime/build/ios/Debug/_deps/abseil_cpp-build/absl/log/Debug-iphonesimulator/libabsl_log_internal_check_op.a /Users/edch2/source/onnxruntime/build/ios/Debug/_deps/abseil_cpp-build/absl/log/Debug-iphonesimulator/libabsl_log_internal_message.a /Users/edch2/source/onnxruntime/build/ios/Debug/_deps/abseil_cpp-build/absl/log/Debug-iphonesimulator/libabsl_log_internal_format.a /Users/edch2/source/onnxruntime/build/ios/Debug/_deps/abseil_cpp-build/absl/strings/Debug-iphonesimulator/libabsl_str_format_internal.a /Users/edch2/source/onnxruntime/build/ios/Debug/_deps/abseil_cpp-build/absl/log/Debug-iphonesimulator/libabsl_log_internal_log_sink_set.a /Users/edch2/source/onnxruntime/build/ios/Debug/_deps/abseil_cpp-build/absl/log/Debug-iphonesimulator/libabsl_log_internal_globals.a /Users/edch2/source/onnxruntime/build/ios/Debug/_deps/abseil_cpp-build/absl/log/Debug-iphonesimulator/libabsl_log_sink.a /Users/edch2/source/onnxruntime/build/ios/Debug/_deps/abseil_cpp-build/absl/log/Debug-iphonesimulator/libabsl_log_entry.a /Users/edch2/source/onnxruntime/build/ios/Debug/_deps/abseil_cpp-build/absl/log/Debug-iphonesimulator/libabsl_log_globals.a /Users/edch2/source/onnxruntime/build/ios/Debug/_deps/abseil_cpp-build/absl/hash/Debug-iphonesimulator/libabsl_hash.a /Users/edch2/source/onnxruntime/build/ios/Debug/_deps/abseil_cpp-build/absl/hash/Debug-iphonesimulator/libabsl_city.a /Users/edch2/source/onnxruntime/build/ios/Debug/_deps/abseil_cpp-build/absl/hash/Debug-iphonesimulator/libabsl_low_level_hash.a /Users/edch2/source/onnxruntime/build/ios/Debug/_deps/abseil_cpp-build/absl/types/Debug-iphonesimulator/libabsl_bad_variant_access.a /Users/edch2/source/onnxruntime/build/ios/Debug/_deps/abseil_cpp-build/absl/log/Debug-iphonesimulator/libabsl_vlog_config_internal.a /Users/edch2/source/onnxruntime/build/ios/Debug/_deps/abseil_cpp-build/absl/synchronization/Debug-iphonesimulator/libabsl_synchronization.a /Users/edch2/source/onnxruntime/build/ios/Debug/_deps/abseil_cpp-build/absl/synchronization/Debug-iphonesimulator/libabsl_kernel_timeout_internal.a /Users/edch2/source/onnxruntime/build/ios/Debug/_deps/abseil_cpp-build/absl/time/Debug-iphonesimulator/libabsl_time.a /Users/edch2/source/onnxruntime/build/ios/Debug/_deps/abseil_cpp-build/absl/time/Debug-iphonesimulator/libabsl_time_zone.a /Users/edch2/source/onnxruntime/build/ios/Debug/_deps/abseil_cpp-build/absl/time/Debug-iphonesimulator/libabsl_civil_time.a /Users/edch2/source/onnxruntime/build/ios/Debug/_deps/abseil_cpp-build/absl/synchronization/Debug-iphonesimulator/libabsl_graphcycles_internal.a /Users/edch2/source/onnxruntime/build/ios/Debug/_deps/abseil_cpp-build/absl/types/Debug-iphonesimulator/libabsl_bad_optional_access.a /Users/edch2/source/onnxruntime/build/ios/Debug/_deps/abseil_cpp-build/absl/log/Debug-iphonesimulator/libabsl_log_internal_fnmatch.a /Users/edch2/source/onnxruntime/build/ios/Debug/_deps/abseil_cpp-build/absl/debugging/Debug-iphonesimulator/libabsl_examine_stack.a /Users/edch2/source/onnxruntime/build/ios/Debug/_deps/abseil_cpp-build/absl/debugging/Debug-iphonesimulator/libabsl_symbolize.a /Users/edch2/source/onnxruntime/build/ios/Debug/_deps/abseil_cpp-build/absl/base/Debug-iphonesimulator/libabsl_malloc_internal.a /Users/edch2/source/onnxruntime/build/ios/Debug/_deps/abseil_cpp-build/absl/debugging/Debug-iphonesimulator/libabsl_demangle_internal.a /Users/edch2/source/onnxruntime/build/ios/Debug/_deps/abseil_cpp-build/absl/debugging/Debug-iphonesimulator/libabsl_demangle_rust.a /Users/edch2/source/onnxruntime/build/ios/Debug/_deps/abseil_cpp-build/absl/debugging/Debug-iphonesimulator/libabsl_decode_rust_punycode.a /Users/edch2/source/onnxruntime/build/ios/Debug/_deps/abseil_cpp-build/absl/debugging/Debug-iphonesimulator/libabsl_utf8_for_code_point.a /Users/edch2/source/onnxruntime/build/ios/Debug/_deps/abseil_cpp-build/absl/debugging/Debug-iphonesimulator/libabsl_stacktrace.a /Users/edch2/source/onnxruntime/build/ios/Debug/_deps/abseil_cpp-build/absl/debugging/Debug-iphonesimulator/libabsl_debugging_internal.a /Users/edch2/source/onnxruntime/build/ios/Debug/_deps/abseil_cpp-build/absl/log/Debug-iphonesimulator/libabsl_log_internal_proto.a /Users/edch2/source/onnxruntime/build/ios/Debug/_deps/abseil_cpp-build/absl/base/Debug-iphonesimulator/libabsl_strerror.a /Users/edch2/source/onnxruntime/build/ios/Debug/_deps/abseil_cpp-build/absl/log/Debug-iphonesimulator/libabsl_log_internal_nullguard.a /Users/edch2/source/onnxruntime/build/ios/Debug/_deps/abseil_cpp-build/absl/strings/Debug-iphonesimulator/libabsl_strings.a /Users/edch2/source/onnxruntime/build/ios/Debug/_deps/abseil_cpp-build/absl/strings/Debug-iphonesimulator/libabsl_strings_internal.a /Users/edch2/source/onnxruntime/build/ios/Debug/_deps/abseil_cpp-build/absl/numeric/Debug-iphonesimulator/libabsl_int128.a /Users/edch2/source/onnxruntime/build/ios/Debug/_deps/abseil_cpp-build/absl/strings/Debug-iphonesimulator/libabsl_string_view.a /Users/edch2/source/onnxruntime/build/ios/Debug/_deps/abseil_cpp-build/absl/base/Debug-iphonesimulator/libabsl_base.a /Users/edch2/source/onnxruntime/build/ios/Debug/_deps/abseil_cpp-build/absl/base/Debug-iphonesimulator/libabsl_spinlock_wait.a /Users/edch2/source/onnxruntime/build/ios/Debug/_deps/abseil_cpp-build/absl/base/Debug-iphonesimulator/libabsl_throw_delegate.a /Users/edch2/source/onnxruntime/build/ios/Debug/_deps/abseil_cpp-build/absl/base/Debug-iphonesimulator/libabsl_raw_logging_internal.a /Users/edch2/source/onnxruntime/build/ios/Debug/_deps/abseil_cpp-build/absl/base/Debug-iphonesimulator/libabsl_log_severity.a /Applications/Xcode.app/Contents/Developer/Platforms/iPhoneSimulator.platform/Developer/SDKs/iPhoneSimulator18.0.sdk/usr/lib/libiconv.tbd -framework XCTest -lXCTestSwiftSupport -Xlinker -dependency_info -Xlinker /Users/edch2/source/onnxruntime/build/ios/Debug/build/onnxruntime_test_all_xc.build/Debug-iphonesimulator/Objects-normal/x86_64/onnxruntime_test_all_xc_dependency_info.dat -o /Users/edch2/source/onnxruntime/build/ios/Debug/Debug-iphonesimulator/onnxruntime_test_all.app/PlugIns/onnxruntime_test_all_xc.xctest/onnxruntime_test_all_xc
```

Reply:
ah, I got it to show up in a CI build:

```
2024-10-10T04:00:26.6922750Z Running ios_package_test failed. Return code was 65
2024-10-10T04:00:26.6923050Z xcrun xcodebuild test stderr:
2024-10-10T04:00:26.6923420Z --- xcodebuild: WARNING: Using the first of multiple matching destinations:
2024-10-10T04:00:26.6924090Z { platform:iOS Simulator, id:BDB8918F-5FB6-4343-B666-EA87BA2CD4D3, OS:18.1, name:iPhone 16 Pro }
2024-10-10T04:00:26.6924820Z { platform:iOS Simulator, id:BDB8918F-5FB6-4343-B666-EA87BA2CD4D3, OS:18.1, name:iPhone 16 Pro }
2024-10-10T04:00:26.6925240Z Testing failed:
2024-10-10T04:00:26.6925960Z 	Undefined symbol: std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>>* absl::log_internal::MakeCheckOpString<char const*, char const*>(char const*, char const*, char const*)
2024-10-10T04:00:26.6927220Z 	Undefined symbol: std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>>* absl::log_internal::MakeCheckOpString<long long, long long>(long long, long long, char const*)
2024-10-10T04:00:26.6928490Z 	Undefined symbol: std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>>* absl::log_internal::MakeCheckOpString<unsigned long long, unsigned long long>(unsigned long long, unsigned long long, char const*)
2024-10-10T04:00:26.6929490Z 	Undefined symbol: absl::log_internal::CheckOpMessageBuilder::ForVar2()
2024-10-10T04:00:26.6929960Z 	Undefined symbol: absl::log_internal::CheckOpMessageBuilder::NewString()
2024-10-10T04:00:26.6930520Z 	Undefined symbol: absl::log_internal::CheckOpMessageBuilder::CheckOpMessageBuilder(char const*)
2024-10-10T04:00:26.6931260Z 	Undefined symbol: absl::log_internal::MakeCheckOpValueString(std::__1::basic_ostream<char, std::__1::char_traits<char>>&, unsigned char)
2024-10-10T04:00:26.6931850Z 	Testing cancelled because the build failed.
2024-10-10T04:00:26.6932040Z 
2024-10-10T04:00:26.6932100Z ** TEST FAILED **
```

https://github.com/microsoft/onnxruntime/actions/runs/11266799368/job/31330892047?pr=22379

Reply:
> I'm not sure why I was seeing the errors. possibly from some intermediate commit. anyway, I can't reproduce the error after reverting those changes so I'll consider them unnecessary.

sorry, I think I was trying to build the wrong test. it's showing up in ios_package_test, not the ORT unit tests.

Reply:
The missing lib is part of ABSL's check library. So, we indirectly used this library? The ArrayFeatureExtractor class indirectly included ABSL's check.h ?

Reply:
We probably should add "absl::check" instead.

Reply:
I think I found the issue. it's not related to ABSEIL_LIBS. when we are constructing the static lib dependencies for the static framework, I think we should be checking the INTERFACE_LINK_LIBRARIES property instead of LINK_LIBRARIES.

checking both just to be safe. it did work with just INTERFACE_LINK_LIBRARIES.